File: Backend/routers/barberos.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, status
from sqlalchemy.orm import Session
from typing import List
from pathlib import Path
import shutil
import time

from database import get_db
import crud.barbero as crud_barbero

from models import LoginBarbero
from schemas import (
    BarberoCreate,
    BarberoUpdate,
    BarberoOut,
    AgendaBarberoOut,
    CrearCuentaBarberoIn
)

from core.dependencias import get_current_login_barbero
from core.security import hash_password


router = APIRouter(
    prefix="/barberos",
    tags=["Barberos"]
)

MEDIA_DIR = Path("static/barberos")


# ---------------------------------------------------------
# AGENDA BARBERO
# ---------------------------------------------------------

@router.get("/mi-agenda", response_model=List[AgendaBarberoOut])
def mi_agenda(
    login=Depends(get_current_login_barbero),
    db: Session = Depends(get_db),
):
    if login.role != "barbero" or not login.barbero_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Este usuario no tiene agenda"
        )

    return crud_barbero.get_agenda_barbero(
        db,
        barbero_id=login.barbero_id
    )


# ---------------------------------------------------------
# PÚBLICO
# ---------------------------------------------------------

@router.get("/", response_model=List[BarberoOut])
def listar_barberos(db: Session = Depends(get_db)):
    return crud_barbero.get_barberos(db)


@router.get("/activos", response_model=List[BarberoOut])
def listar_barberos_activos(db: Session = Depends(get_db)):
    return crud_barbero.get_barberos(db, solo_activos=True)


@router.get("/{barbero_id}", response_model=BarberoOut)
def obtener_barbero(barbero_id: int, db: Session = Depends(get_db)):
    barbero = crud_barbero.get_barbero_by_id(db, barbero_id)
    if not barbero:
        raise HTTPException(404, "Barbero no encontrado")

    return crud_barbero.serialize_barbero(barbero)


# ---------------------------------------------------------
# ADMIN
# ---------------------------------------------------------

@router.post("/", response_model=BarberoOut, status_code=status.HTTP_201_CREATED)
def crear_barbero(
    barbero_in: BarberoCreate,
    db: Session = Depends(get_db),
):
    barbero = crud_barbero.create_barbero(db, barbero_in)
    return crud_barbero.serialize_barbero(barbero)


@router.put("/{barbero_id}", response_model=BarberoOut)
def actualizar_barbero(
    barbero_id: int,
    barbero_in: BarberoUpdate,
    db: Session = Depends(get_db),
):
    barbero = crud_barbero.get_barbero_by_id(db, barbero_id)
    if not barbero:
        raise HTTPException(404, "Barbero no encontrado")

    barbero = crud_barbero.update_barbero(db, barbero, barbero_in)
    return crud_barbero.serialize_barbero(barbero)


@router.patch("/{barbero_id}/toggle", response_model=BarberoOut)
def toggle_barbero(barbero_id: int, db: Session = Depends(get_db)):
    barbero = crud_barbero.get_barbero_by_id(db, barbero_id)
    if not barbero:
        raise HTTPException(404, "Barbero no encontrado")

    barbero = crud_barbero.toggle_barbero_estado(db, barbero)
    return crud_barbero.serialize_barbero(barbero)


@router.delete("/{barbero_id}", status_code=status.HTTP_204_NO_CONTENT)
def eliminar_barbero(barbero_id: int, db: Session = Depends(get_db)):
    barbero = crud_barbero.get_barbero_by_id(db, barbero_id)
    if not barbero:
        raise HTTPException(404, "Barbero no encontrado")

    try:
        crud_barbero.delete_barbero(db, barbero)
    except ValueError as e:
        if str(e) == "BARBERO_CON_VISITAS":
            raise HTTPException(
                status_code=400,
                detail="No se puede eliminar un barbero con visitas asociadas"
            )
        raise

    return None


# ---------------------------------------------------------
# CREAR ACCESO PARA BARBERO
# ---------------------------------------------------------

@router.post("/{barbero_id}/crear-acceso", status_code=status.HTTP_201_CREATED)
def crear_acceso_barbero(
    barbero_id: int,
    data: CrearCuentaBarberoIn,
    db: Session = Depends(get_db),
    admin=Depends(get_current_login_barbero),
):
    if admin.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="No autorizado"
        )

    barbero = crud_barbero.get_barbero_by_id(db, barbero_id)
    if not barbero:
        raise HTTPException(404, "Barbero no encontrado")

    existe = (
        db.query(LoginBarbero)
        .filter(LoginBarbero.barbero_id == barbero_id)
        .first()
    )
    if existe:
        raise HTTPException(
            status_code=400,
            detail="El barbero ya tiene acceso"
        )

    login = LoginBarbero(
        nombre=barbero.nombre,
        email=data.email,
        password_hash=hash_password(data.password),
        role=data.rol,
        barbero_id=barbero_id,
        is_active=True
    )

    db.add(login)
    db.commit()
    db.refresh(login)

    return {"ok": True}


# ---------------------------------------------------------
# SUBIR FOTO BARBERO
# ---------------------------------------------------------

@router.post("/{barbero_id}/foto", response_model=BarberoOut) 
def subir_foto_barbero(
    barbero_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
):
    barbero = crud_barbero.get_barbero_by_id(db, barbero_id)
    if not barbero:
        raise HTTPException(404, "Barbero no encontrado")

    MEDIA_DIR.mkdir(parents=True, exist_ok=True)
    
    # 1. Borrar foto anterior (opcional para ahorrar espacio)
    if barbero.foto_url and barbero.foto_url != "default.jpg":
        old_path = MEDIA_DIR / barbero.foto_url
        if old_path.exists():
            old_path.unlink()

    # 2. Guardar nueva foto
    timestamp = int(time.time())
    ext = Path(file.filename).suffix or ".jpg"
    filename = f"barbero_{barbero_id}_{timestamp}{ext}" 
    filepath = MEDIA_DIR / filename

    with open(filepath, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)

    barbero.foto_url = filename
    db.commit()

    return crud_barbero.serialize_barbero(barbero)

File: Frontend/Barberia/src/services/barberos.js
import API_URL from "./api";

// =======================================================
// BARBEROS
// =======================================================

export async function getBarberos() {
  const res = await fetch(`${API_URL}/barberos/`);

  if (!res.ok) {
    throw new Error("Error al cargar barberos");
  }

  return await res.json();
}

export async function crearBarbero(data) {
  const res = await fetch(`${API_URL}/barberos/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
    body: JSON.stringify(data),
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail || "Error al crear barbero");
  }

  return await res.json();
}

export async function subirFotoBarbero(idBarbero, file) {
  if (!idBarbero) {
    throw new Error("ID de barbero inválido");
  }

  const formData = new FormData();
  formData.append("file", file);

  const res = await fetch(`${API_URL}/barberos/${idBarbero}/foto`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
    body: formData,
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail || "Error al subir foto");
  }

  return await res.json();
}

export async function toggleBarbero(idBarbero) {
  if (!idBarbero) {
    throw new Error("ID de barbero inválido");
  }

  const res = await fetch(`${API_URL}/barberos/${idBarbero}/toggle`, {
    method: "PATCH",
    headers: {
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(
      err.detail || "Error al cambiar estado del barbero"
    );
  }

  return await res.json();
}

export async function eliminarBarbero(idBarbero) {
  if (!idBarbero) {
    throw new Error("ID de barbero inválido");
  }

  const res = await fetch(`${API_URL}/barberos/${idBarbero}`, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${localStorage.getItem("token")}`,
    },
  });

  if (!res.ok) {
    const err = await res.json();
    throw new Error(err.detail || "Error al eliminar barbero");
  }

  return res.status === 204 ? true : await res.json();
}

// =======================================================
// AGENDA DEL BARBERO (CON AUTH + FILTRO POR FECHA)
// =======================================================

export async function getAgendaBarbero(fecha = null) {
  const query = fecha ? `?fecha=${fecha}` : "";

  const res = await fetch(
    `${API_URL}/visitas/mi-agenda${query}`,
    {
      headers: {
        Authorization: `Bearer ${localStorage.getItem("token")}`,
      },
    }
  );

  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || "Error al cargar agenda del barbero");
  }

  return await res.json();
}

// =======================================================
// CREAR CUENTA (LOGIN) PARA BARBERO — ADMIN
// =======================================================

/**
 * @param {number} barberoId  -> ID del barbero (VA EN LA URL)
 * @param {object} data       -> { nombre, email, password, rol }
 */
export async function crearCuentaBarbero(barberoId, data) {
  if (!barberoId) {
    throw new Error("ID de barbero inválido");
  }

  const res = await fetch(
    `${API_URL}/barberos/${barberoId}/crear-acceso`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${localStorage.getItem("token")}`,
      },
      body: JSON.stringify(data),
    }
  );

  if (!res.ok) {
    const err = await res.json();
    throw new Error(
      err.detail || "Error al crear la cuenta del barbero"
    );
  }

  return await res.json();
}
